File: farebot-transit-octopus/src/main/java/com/codebutler/farebot/transit/octopus/OctopusTransitFactory.java
/*
 * OctopusTransitFactory.java
 *
 * Copyright 2016 Michael Farrell <micolous+git@gmail.com>
 *
 * Portions based on FelicaCard.java from nfcard project
 * Copyright 2013 Sinpo Wei <sinpowei@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.transit.octopus;

import android.support.annotation.NonNull;
import com.codebutler.farebot.base.util.ByteArray;
import com.codebutler.farebot.base.util.ByteUtils;
import com.codebutler.farebot.card.felica.FelicaCard;
import com.codebutler.farebot.card.felica.FelicaService;
import com.codebutler.farebot.card.felica.FelicaSystem;
import com.codebutler.farebot.transit.TransitFactory;
import com.codebutler.farebot.transit.TransitIdentity;
import com.codebutler.farebot.transit.registry.annotations.TransitCard;
import net.kazzz.felica.lib.FeliCaLib;

import static com.codebutler.farebot.transit.octopus.OctopusTransitInfo.*;

@TransitCard
public class OctopusTransitFactory implements TransitFactory<FelicaCard, OctopusTransitInfo> {

    @Override
    public boolean check(@NonNull FelicaCard card) {
        return (card.getSystem(FeliCaLib.SYSTEMCODE_OCTOPUS) != null)
                || (card.getSystem(FeliCaLib.SYSTEMCODE_SZT) != null);
    }

    @NonNull
    @Override
    public TransitIdentity parseIdentity(@NonNull FelicaCard card) {
        if (card.getSystem(FeliCaLib.SYSTEMCODE_SZT) != null) {
            if (card.getSystem(FeliCaLib.SYSTEMCODE_OCTOPUS) != null) {
                // Dual-mode card.
                return TransitIdentity.create(DUAL_NAME, null);
            } else {
                // SZT-only card.
                return TransitIdentity.create(SZT_NAME, null);
            }
        } else {
            // Octopus-only card.
            return TransitIdentity.create(OCTOPUS_NAME, null);
        }
    }

    @NonNull
    @Override
    public OctopusTransitInfo parseInfo(@NonNull FelicaCard card) {
        int octopusBalance = 0;
        int shenzhenBalance = 0;
        boolean hasOctopus = false;
        boolean hasShenzhen = false;

        FelicaSystem octopusSystem = card.getSystem(FeliCaLib.SYSTEMCODE_OCTOPUS);
        if (octopusSystem != null) {
            FelicaService service = octopusSystem.getService(FeliCaLib.SERVICE_OCTOPUS);
            if (service != null) {
                ByteArray metadata = service.getBlocks().get(0).getData();
                octopusBalance = ByteUtils.byteArrayToInt(metadata.bytes(), 0, 4) - 350;
                hasOctopus = true;
            }
        }

        FelicaSystem sztSystem = card.getSystem(FeliCaLib.SYSTEMCODE_SZT);
        if (sztSystem != null) {
            FelicaService service = sztSystem.getService(FeliCaLib.SERVICE_SZT);
            if (service != null) {
                ByteArray metadata = service.getBlocks().get(0).getData();
                shenzhenBalance = ByteUtils.byteArrayToInt(metadata.bytes(), 0, 4) - 350;
                hasShenzhen = true;
            }
        }

        return OctopusTransitInfo.create(octopusBalance, shenzhenBalance, hasOctopus, hasShenzhen);
    }
}

File: farebot-transit-octopus/src/commonMain/kotlin/com/codebutler/farebot/transit/octopus/OctopusTransitFactory.kt
/*
 * OctopusTransitFactory.kt
 *
 * Copyright 2016 Michael Farrell <micolous+git@gmail.com>
 *
 * Portions based on FelicaCard.java from nfcard project
 * Copyright 2013 Sinpo Wei <sinpowei@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.transit.octopus

import com.codebutler.farebot.base.util.ByteUtils
import com.codebutler.farebot.card.CardType
import com.codebutler.farebot.card.felica.FelicaCard
import com.codebutler.farebot.card.felica.FeliCaConstants
import com.codebutler.farebot.transit.CardInfo
import com.codebutler.farebot.transit.TransitFactory
import com.codebutler.farebot.transit.TransitIdentity
import com.codebutler.farebot.transit.TransitRegion
import farebot.farebot_transit_octopus.generated.resources.*
import kotlinx.coroutines.runBlocking
import org.jetbrains.compose.resources.getString

class OctopusTransitFactory : TransitFactory<FelicaCard, OctopusTransitInfo> {

    override val allCards: List<CardInfo>
        get() = listOf(CARD_INFO)

    override fun check(card: FelicaCard): Boolean {
        return (card.getSystem(FeliCaConstants.SYSTEMCODE_OCTOPUS) != null)
                || (card.getSystem(FeliCaConstants.SYSTEMCODE_SZT) != null)
    }

    override fun parseIdentity(card: FelicaCard): TransitIdentity = runBlocking {
        if (card.getSystem(FeliCaConstants.SYSTEMCODE_SZT) != null) {
            if (card.getSystem(FeliCaConstants.SYSTEMCODE_OCTOPUS) != null) {
                // Dual-mode card.
                TransitIdentity.create(getString(Res.string.octopus_dual_card_name), null)
            } else {
                // SZT-only card.
                TransitIdentity.create(getString(Res.string.octopus_szt_card_name), null)
            }
        } else {
            // Octopus-only card.
            TransitIdentity.create(getString(Res.string.octopus_card_name), null)
        }
    }

    override fun parseInfo(card: FelicaCard): OctopusTransitInfo {
        var octopusBalance: Int? = null
        var shenzhenBalance: Int? = null

        val octopusSystem = card.getSystem(FeliCaConstants.SYSTEMCODE_OCTOPUS)
        if (octopusSystem != null) {
            val service = octopusSystem.getService(FeliCaConstants.SERVICE_OCTOPUS)
            if (service != null) {
                val metadata = service.blocks[0].data
                val rawBalance = ByteUtils.byteArrayToInt(metadata, 0, 4)
                // Apply date-dependent offset and convert from 10-cent units to cents
                octopusBalance = (rawBalance - OctopusData.getOctopusOffset(card.scannedAt)) * 10
            }
        }

        val sztSystem = card.getSystem(FeliCaConstants.SYSTEMCODE_SZT)
        if (sztSystem != null) {
            val service = sztSystem.getService(FeliCaConstants.SERVICE_SZT)
            if (service != null) {
                val metadata = service.blocks[0].data
                val rawBalance = ByteUtils.byteArrayToInt(metadata, 0, 4)
                // Apply offset and convert from 10-cent units to cents
                shenzhenBalance = (rawBalance - OctopusData.getShenzhenOffset(card.scannedAt)) * 10
            }
        }

        return OctopusTransitInfo.create(
            octopusBalance,
            shenzhenBalance,
            hasOctopus = octopusBalance != null,
            hasShenzen = shenzhenBalance != null
        )
    }

    companion object {
        private val CARD_INFO = CardInfo(
            name = runBlocking { getString(Res.string.octopus_card_name) },
            cardType = CardType.FeliCa,
            region = TransitRegion.HONG_KONG,
            locationId = Res.string.location_hong_kong
        )
    }
}
